praxis-isncsci/ui
# Simple grid view to show raw data

Closes #15 

## Problem

Now that we have an External Port provider, we need to test that we can pass data back and forth. To visually confirm data is coming in, we need a basic grid view to show the raw data.

## Solution

- [x] Initialize state gridModel - Add `src/core/useCases/initializeApp.useCase.ts` and code to initialize the app and set an empty grid model as default
- [x] Update grid with model values - Adds `src/app/controllers/inputLayout.controller.ts` and code needed to update the grid view with the grid model data
- [x] Validate exam data and bind to grid model - Adds helper code and `src/core/useCases/loadExternalExamData.useCase.ts` file which validates incoming exam data and binds it to the grid model
- [x] Update state on external exam data - Has the updates to the app store provider and state objects to update the grid model and views when external exam data is passed
- [x] Add jest and unit tests for validateExamData
- [x] Add GitHub action `Jest Unit Tests`

## Test

- [x] 1. Can update state and view with external exam data

<details>
  <summary>Test case</summary>

  1. Navigate to [Chromatic story](https://64f8d7c6e093108e99084a70-ggnyajelfe.chromatic.com/?path=/docs/app-externalmessageportprovider--docs)
  2. Scroll down to the `Primary` story
  3. Press the `Load random exam` button
  4. Confirm that exam data loads on the grids

  By reviewing the code you should see that random exam data is generated by the story and passed to the app via the `MessageChannel` port. After receiving it, the app calls `loadExternalExamDataUseCase` which binds the incoming data to a grid mode and updates the state with the new model. The new controller updates the UI with the new state.

  ![image](https://github.com/praxis-isncsci/ui/assets/1294355/1dd7ee0b-a71e-4b90-9797-fdd2ed3ca94a)

</details>

---

- [x] 2. Unit tests run on commit

<details>
  <summary>Test case</summary>

  1. Scroll down on this PR to the `Checks` section
  2. Confirm the `Jest Unit Tests / test (push)` action ran
  3. Navigate to the [Details](https://github.com/praxis-isncsci/ui/actions/runs/6152599333/job/16695025822?pr=18) and expand the `Run the tests` section
  4. Confirm that the tests were executed

  By reviewing the code you should see that random exam data is generated by the story and passed to the app via the `MessageChannel` port. After receiving it, the app calls `loadExternalExamDataUseCase` which binds the incoming data to a grid mode and updates the state with the new model. The new controller updates the UI with the new state.

  ![image](https://github.com/praxis-isncsci/ui/assets/1294355/5fa48cd3-cc3b-491a-922f-1ca6811ac10b)

  ![image](https://github.com/praxis-isncsci/ui/assets/1294355/ac01d8bf-8d5e-4cc7-81e5-5f20be9cb386)

</details>

---
